File: Python/4Sum.py
class Solution:
    # @return a list of lists of length 4, [[val1,val2,val3,val4]]
    def fourSum(self, num, target):
        res = []
        if len(num) < 4:
            return res

        num.sort()

        for i in range(len(num) -2):
            for v in range(i+1, len(num)-1):
                j = v + 1
                k = len(num) -1
                while j<k :
                    cur = num[i] + num[j] + num[k] + num[v]
                    if cur == target and not [num[i], num[v], num[j],num[k]] in res:
                        res.append([num[i], num[v], num[j],num[k]])
                        k = k-1;
                        j = j+1;
                        while (j<k-1 and num[k] == num[k+1]):
                            k = k-1
                        while (j<k-1 and num[j] == num[j-1]):
                            j = j+1
      
                    elif cur > target:
                        k = k - 1
                    else:
                        j = j + 1

        return res

File: Java/src/4Sum.java
public class Solution {
    public List<List<Integer>> fourSum(int[] num, int target) {
        List<List<Integer>> ret = new ArrayList<>();

        if (num == null || num.length < 4) return ret;
        Arrays.sort(num);

        for (int v=0; v < num.length-2; v++){
            for(int i=v+1; i< num.length-1; i++){
                int left = i+1;
                int right = num.length -1;
                while (left < right){
                    int sum = num[v]+ num[i] + num[left] + num[right];
                    if (sum == target){
                        List<Integer> result = Arrays.asList(num[v], num[i], num[left], num[right]);
                        if (!ret.contains(result)) ret.add(result);
                        left++;
                        right--;
                        while(left<right&&num[left]==num[left-1]){
                            left++;
                        }
                        while(left<right&&num[right]==num[right+1]){
                            right--;
                        }
                    }
                    else if(sum < target){
                        left++;
                    }
                    else{
                        right--;
                    }
                }
            }
        }
        return ret;
    }
}